apps/bgu/bgu_generator.cpp

```diff
@@ -126,6 +127,86 @@ Matrix<M, N> solve(Matrix<M, M> A, Matrix<M, N> b, Func compute, Var at, bool sk
     return b;
 };
 
+// Solve Ax = b at each x, y, z exploiting the fact that A is
+// symmetric. Compute the result at the given Func and Var.
+template<int M, int N>
+Matrix<M, N> solve_symmetric(Matrix<M, M> A, Matrix<M, N> b,
+                             Func compute, Var at, bool skip_schedule, Target target) {
+
+    // Put the input matrices in a Func to do sqrt-free Cholesky.
+    Var vi, vj;
+    Func f;
+    f(x, y, z, vi, vj) = undef<float>();
+    for (int i = 0; i < M; i++) {
+        for (int j = 0; j < M; j++) {
+            f(x, y, z, i, j) = A(i, j);
+        }
+        for (int j = 0; j < N; j++) {
+            f(x, y, z, i, j + M) = b(i, j);
+        }
+    }
+
+    for (int j = 0; j < M; j++) {
+        // Normalize the jth column starting at the jth row, storing
+        // the normalization factor on the diagonal. The unnormalized
+        // version stays in the jth row.
```

Comment:
Is it faster / cheaper to store everything in place in Halide? It's more readable to have separate L and D Funcs.

Reply:
Yeah keeping it all in update definitions in the one Func makes the mutual recursion legal.